src/btree/bt_random.c

```diff
@@ -9,147 +9,332 @@
 #include "wt_internal.h"
 
 /*
- * __wt_row_random_leaf --
- *     Return a random key from a row-store leaf page.
+ * __random_insert_valid --
+ *     Check if the inserted key/value pair is valid.
  */
-int
-__wt_row_random_leaf(WT_SESSION_IMPL *session, WT_CURSOR_BTREE *cbt)
+static int
+__random_insert_valid(
+  WT_CURSOR_BTREE *cbt, WT_INSERT_HEAD *ins_head, WT_INSERT *ins, WT_UPDATE **updp, bool *validp)
 {
-    WT_INSERT *ins, **start, **stop;
-    WT_INSERT_HEAD *ins_head;
-    WT_PAGE *page;
-    uint64_t samples;
-    uint32_t choice, entries, i;
-    int level;
+    *updp = NULL;
+    *validp = false;
 
-    page = cbt->ref->page;
-    start = stop = NULL; /* [-Wconditional-uninitialized] */
-    entries = 0;         /* [-Wconditional-uninitialized] */
+    __cursor_pos_clear(cbt);
+    cbt->slot = 0;
+    cbt->ins_head = ins_head;
+    cbt->ins = ins;
+    cbt->compare = 0;
+
+    return (__wt_cursor_valid(cbt, updp, validp));
+}
+
+/*
+ * __random_slot_valid --
+ *     Check if the slot key/value pair is valid.
+ */
+static int
+__random_slot_valid(WT_CURSOR_BTREE *cbt, uint32_t slot, WT_UPDATE **updp, bool *validp)
+{
+    *updp = NULL;
+    *validp = false;
 
     __cursor_pos_clear(cbt);
+    cbt->slot = slot;
+    cbt->compare = 0;
+
+    return (__wt_cursor_valid(cbt, updp, validp));
+}
 
-    /* If the page has disk-based entries, select from them. */
-    if (page->entries != 0) {
-        cbt->compare = 0;
-        cbt->slot = __wt_random(&session->rnd) % page->entries;
+/* Magic constant: 50 entries in a skip list is enough to predict the size. */
+#define WT_RANDOM_SKIP_PREDICT 50
+/* Magic constant: 5000 entries in a skip list is enough to forcibly evict. */
+#define WT_RANDOM_SKIP_EVICT_SOON 5000
```

Comment:
Can all of these be moved to the top of the file (or an appropriate header file)? Having them scattered many times throughout the source file, even though I know you put them near the function using them, is difficult to read and since they all have a similar or the same prefix, there's risk of overlapping names since you cannot visually see them together.

Reply:
I'd prefer not to move them (as you note, they appear where they're used), and they're only used once. Overlapping names would get flagged by the compiler, right?

We can give @sulabhM the vote. :)

Reply:
There really isn't much precedent for this form. Similar usage is in `conn/conn_cache_pool.c` has them at the top. In any case, @sulabhM can be the tiebreaker. At a minimum, if they stay, please alphabetize them in their little groups.

Reply:
Sorted in 2611383.

Reply:
I don't like the idea of a separate header file. I don't really ever scroll the file from top to the bottom. I search for a function and scroll from there and it is useful to have related functionality clubbed together. I vote for leaving them as they are today. I have nothing against putting them together at the top, just that it would not help me in the way I search through the code. I believe most others do the same.